waiter/test/waiter/auth/oauth_test.clj

```diff
@@ -0,0 +1,211 @@
+        (is (= 307 status))))
+    (testing "unauthenticated"
+      (let [{{:strs [location]} :headers :keys [status]}
+            (wrapped-handler {:uri "/requires-auth"})]
+        (is (= "/waiter-auth/oauth/providers/requires-auth" location))
+        (is (= 307 status))))
+    (testing "already authenticated"
+      (let [cookie (str auth/AUTH-COOKIE-NAME "=" (-> (cs/encode-cookie [principal now] password)
+                                                      UrlEncoded/encodeString))
+            {{:strs [location]} :headers :keys [body status]}
+            (wrapped-handler {:uri "/requires-auth"
+                              :headers {"cookie" cookie}})]
+        (is (= 200 status))
+        (is (= body "handler"))))))
+
+(deftest test-github-provider
+  (let [authenticate-uri-fn (constantly "http://authenticate")
+        client-id "client-id"
+        client-secret "client-secret"
+        password [:cached "password"]
+        provider (map->GitHubOAuthProvider {:authenticate-uri-fn authenticate-uri-fn
+                                            :client-id client-id
+                                            :client-secret client-secret
+                                            :password password})
+        token "abc123"
+        state {:location "/path?a=1"
+               :token token}
+        cookie (str OAUTH-COOKIE-NAME "=" (-> (cs/encode-cookie token password)
+                                              UrlEncoded/encodeString))]
+    (testing "redirect"
+      (let [{{:strs [set-cookie location]} :headers :keys [status]} (fa/<?? (redirect provider {}))]
+        (is location)
+        (is (= status 307))
+        (is set-cookie)))
+    (testing "authenticate"
+      (with-redefs [random-str (fn [] token)
+                    http/post (fn [_ _ _]
+                                (async/go {:body (async/go "access_token=accesstoken&token_type=bearer")
+                                           :status 200}))
+                    http/get (fn [_ _ _]
+                               (async/go
+                                 {:body (async/go
+                                          (json/write-str [{:email "user@example.com"
+                                                            :primary true
+                                                            :verified true}]))
+                                  :status 200}))]
+        (let [{{:strs [location]} :headers :keys [status] :as response}
+              (fa/<?? (authenticate provider {:headers {"cookie" cookie}
+                                              :query-string (str "code=code&state="
+                                                                 (-> state
+                                                                     json/write-str
+                                                                     UrlEncoded/encodeString))}))]
+          (is location)
+          (is (= status 307)))))))
+
+(deftest test-make-uri
```

Comment:
We should also include a test where q query parameter is repeated, e.g. `{a [1 2]}`.

Reply:
It's never used this way.

Reply:
Can we clarify this in the docstring then?